File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int [] result = new int[2];
        int sum =0;
        //Brute Force
        // for(int i=0;i<nums.length;i++){
        //     for(int j=i+1;j<nums.length;j++){
        //         sum=nums[i]+nums[j];
        //         if(sum==target){
        //             result[0]=i;
        //             result[1]=j;
        //         }
        //     }
        // }
        // return result;
        //Optimized : HashMap
        Map<Integer,Integer> map = new HashMap<>();
        for(int i=0;i<nums.length;i++){
            map.put(nums[i],i);
        }

        for(int i=0;i<nums.length;i++){
            int complement = target - nums[i];
            if(map.containsKey(complement) && map.get(complement)!=i){
                return new int[]{i,map.get(complement)};
            }
        }
        return new int[]{};
    }
}

File: 0001-two-sum/0001-two-sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut hmap:HashMap<i32,i32> = HashMap::new();

        for (i,num) in nums.iter().enumerate(){
            let complement = target-num;

            if let Some(&complement_index) = hmap.get(&complement) {
                return vec![complement_index, i as i32];
            }

            hmap.insert(*num, i as i32);
        }

        vec![]
    }
}